riscv/src/compiler.rs

```diff
@@ -695,8 +814,8 @@ fn preamble() -> String {
     // ================== wrapping instructions ==============
 
     // Wraps a value in Y to 32 bits.
-    // Requires 0 <= Y < 2**33
-    instr wrap Y -> X { Y = X + wrap_bit * 2**32, X = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 }
+    // Requires 0 <= Y < 2**64
+    instr wrap Y -> X { Y = X + X_b5 * 2**32 + X_b6 * 2**40 + X_b7 * 2**48 + X_b8 * 2**56, X = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 }
```

Comment:
@chriseth maybe check this too

Reply:
One idea how we could bit-decompose arbitrary field elements. Let $l$ be the number of bits needed to represent one field element, i.e., $2^{l - 1} < p < 2^l$.

Then, the steps are the following:
1. Given a field element $x$, the prover provides $b_i \in \\{0, 1\\}$ s.t. $$x = \sum_{i = 1}^{l}{b_i \cdot 2^{i - 1}} (\mod p)$$
2. We have to make sure that the right-hand side does not overflow, which can only happen if the two most significant bits are equal to the most significant bits of $p$. The most significant bit of $p$ is always $1$ (by definition of $l$). Distinguish the two cases:
    1. If the second most significant bit of $p$ is 0, add the constraint $b_{l - 1} = 0$ and set $check = b_l$
    2. Else, set $check = b_l \cdot b_{l - 1}$
3. If $check = 1$, we ask the prover to provide a diff $d \in \\{0, ..., 2^{l - 2} - 1\\}$ s.t. $$(\sum_{i = 1}^{l - 2}{b_i \cdot 2^{i - 1}}) + d = p \bmod 2^{l - 2}$$
  Note that the left-hand side cannot overflow because we're adding two $(l -2)$-bit numbers, so the result is smaller than $2^{l - 1} < p$. The right-hand side is a constant.

Of course, we don't want to do bit decompositions for efficiency, so this would have to be adjusted further.

Is there a simpler way? I feel like this problem must have been solved before.

Note also that for any concrete field, it could be potentially simplified. For example, for Goldilocks:
- Check that the most significant byte no larger than $128$.
- If it is $128$, check than bytes 5-7 are $0$.

Reply:
Discussed this with @Schaeff and we found a less ad-hoc solution 🎉 Implemented in #570.